[CALCITE-7017] Incorrect subquery rewriting in join conditions when j…

suibianwanwank opened the pull request:
…oin type is outer join

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -926,6 +923,11 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
     if (inputIntersectsLeftSide) {
       builder.push(join.getLeft());
 
+      final RelOptUtil.Logic logic =
```
this code could use a little comment.
(I am not familiar with this part of the code...)

xiedeyantu commented on core/src/test/resources/sql/sub-query.iq:
```diff
@@ -848,6 +848,38 @@ where sal + 100 not in (
+    SELECT e3.empno
+    FROM emp e3
+    WHERE e1.comm > e3.comm
+  )
```
I may not understand the purpose of the test `not in`. Do you want to add a case where `inputIntersectsLeftSide=false`?

asolimando replied:
@suibianwanwank, #4328 has been merged, can you try again?

asolimando commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -946,6 +948,11 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
       builder.push(join.getLeft());
       builder.push(join.getRight());
 
+      final RelOptUtil.Logic logic =
```
The two logic variables in the `if/else` seem identical, in case let's move it to its original location in its updated form

asolimando commented:
> While working on [[CALCITE-7034](https://issues.apache.org/jira/browse/CALCITE-7034)], I encountered some obstacles and might need more time to complete it. I'm inclined to defer it to 1.41.0.

That's totally fine, thanks for letting me know, I will remove the `fixVersion` from the ticket

suibianwanwank commented:
Thank you all for your comments and reviews. I've refactored this PR to address more related issues here. Although they are not directly connected, they were all exposed and affecting each other within the same test. Given that the code changes required are minimal, I'm inclined to resolve them together in this PR.  

For this SQL.

```
SELECT e1.*, d.deptno
FROM emp e1
LEFT JOIN dept d
  ON e1.deptno = d.deptno
  AND d.deptno IN (
    SELECT e3.deptno
    FROM emp e3
    WHERE d.deptno > e3.comm
  );
```
The current code has the following issues:  

   1. This PR addresses the problem described in the Jira ticket, where an outer join produces output even when there are no matching rows.  
   2. The previous handling of the right side of the join was incorrect—it requires not only `shiftFieldAccess` but also shifting `subQuery.getOperands()`.  
   3. In `rewriteIn`, if `variablesSet` is not empty, `builder.join` generates multiple `Correlate` nodes with the same `corId`, which leads to decorrelation failure (I believe this results in an invalid `RelNode`).

mihaibudiu commented:
there is one failure on MacOS
```
 FAILURE  89.6sec, org.apache.calcite.test.CoreQuidemTest > test(String)[25], [25] sql/spatial.iq
    org.opentest4j.AssertionFailedError: Files differ: /Users/runner/work/calcite/calcite/core/build/quidem/test/sql/spatial.iq /Users/runner/work/calcite/calcite/core/build/resources/test/sql/spatial.iq
    1847c1847
    < LINESTRING (0 0, 1 1, 0 0, 1 1, 2 2)
    ---
    > LINESTRING (NaN 0, 1 1, 0 0, 1 1, 2 2)
```

mihaibudiu commented:
The CI failure does not make any sense, maybe the CI machine is broken?

suibianwanwank commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -840,6 +857,29 @@ private static List<RexNode> fields(RelBuilder builder, int fieldCount) {
 
+  /**
+   * Replace every {@link RexFieldAccess} in the given {@link RelNode}
+   * that reference the specified {@link CorrelationId} with another {@link CorrelationId}.
```
I noticed that the type parameter can probably be left out, I'll try to modify.

suibianwanwank commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -960,13 +1001,45 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
+      RexSubQuery subQuery = e;
+
+      if (!variablesSet.isEmpty()) {
+        // Original correlates reference joint row type, but we are about to create
```
I added a plan, but the fieldAccess index wasn't actually printed in the plan. At least showing the plan before and after rewriting helps understand the offset reason.

suibianwanwank commented:
Thank you for reviewing. All issues mentioned in the comments have been resolved.

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -840,6 +857,29 @@ private static List<RexNode> fields(RelBuilder builder, int fieldCount) {
+   */
+  public static RelNode replaceCorrelationId(RexBuilder rexBuilder, RelNode node,
+      final CorrelationId from, final CorrelationId to, RelDataType type) {
+    return node.accept(new RelHomogeneousShuttle() {
```
I am skeptical, because there is no way the RexVisitor can call your homogeneous shuttle back, since it does not have a handle to it.
Can there be subqueries in expressions at this point, or are you assuming they have been expanded?

suibianwanwank replied:
You're right, now I understand what you meant. I've made the corresponding changes, but currently this rule's implementation has other issues in multi-subQuery scenarios. I might need to spend some more time on it.

mihaibudiu commented:
So this is now ready for review?
Should we first merge the well-known count bug fix?
Or are these completely independent?

suibianwanwank commented:
> Should we first merge the well-known count bug fix? Or are these completely independent?

Yes, we should merge the well-known count bug fix first. This fix has no dependencies. 

I want to wait for [CALCITE-7045] to be completed, because the current PR addresses some issues from [CALCITE-7045] to fix the tests (even though this is unrelated to the current issue), but I don't think this is the fundamental solution.
